### DISCUSSION TITLE
Interrupt too slow to count 21Hz?

### DISCUSSION THREAD

**Discussion by pcfreak1201** [earlephilhower/arduino-pico#1629] [Q&A] [answered]
Hi all!

I try to count a sqarewave-signal on GPIO4 with 21-96Hz and use at least 
`attachInterrupt(digitalPinToInterrupt(4), countOne, FALLING);`
and
```
void __not_in_flash_func(countOne)()
{
  uwIN1++;
}
```
... and it counts ... but only one number per second?!
![PIC_0](https://github.com/earlephilhower/arduino-pico/assets/68873801/af2cd767-5a29-44c6-96f2-c3bafdc67853)
Is there a so low limit for detecting IRQs? Of course, I use also ADC and PWM...

Any ideas?

**Comment by earlephilhower** [comment]
For I2S we have run 1000s of interrupts per seconds doing a lot more work than just incrementing a variable.  I suggest making a simple MCVE sketch that just counts that edge and posting it so we can offer some advice.

**Comment by pcfreak1201** [comment]
No problem:
```
#define IN1 4
#define PWM1  2
#define POTI  A2

volatile uint16_t uwIN1 = 0;

void __not_in_flash_func(countOne)()
{
  uwIN1++;
}

void setup(){
    Serial.begin(115200);
    Serial.println("4Fan-Control");

    analogReadResolution(10);   // 10bit ADC-Val
    analogWriteResolution(10);  // 10bit PWM-Res.
    analogWriteFreq(25000); // 25kHz
    analogWriteRange(1000); // 0-1000 "%"

    pinMode(IN1, INPUT);  // ext. 4k7 PullUp to Vcc
    pinMode(PWM1, OUTPUT);  // 25kHz
    attachInterrupt(digitalPinToInterrupt(IN1), countOne, FALLING);
}

void loop(){
  
uint16_t  uiAdcIn = analogRead(POTI);

  analogWrite(PWM1, map(uiAdcIn, 2, 1021, 0, 100));
  Serial.println(uwIN1); // counts per loop
  // uwIN1 = 0; <== counts up, but stays 0 when uncommented
}
```

**Reply by earlephilhower** [reply]
Just looking at it, can you make the global a `uint32_t`?  You're updating a volatile global that's not atomic (16b access are read-modify-write) from two spots so it could get weird.

Otherwise, are you missing a `delay()` in your `loop()`?  I'm not sure how you'd know what frequency you're getting.

Also, how are you driving the counting pin?  Is it possible to drop the pwm frequency to, say, 1khz, and then make a sketch looping back the output PWM to the IRQ pin?

**Comment by pcfreak1201** [comment]
Yes, I already used uint32, no changes.
The absolute frequency isn't the goal. I have to syncronize 4 fans - so I've to do it four times and trying to keep speed when they get static pressure. But the base speed is set with the poti (only once).
The delay is actual the serial.print - but adding delay(10); doesn't change anything.
The input signal is shown in the picture. It comes from the fan, nothing I can change.

So you want me to probe the output signal with the sketch?

**Reply by earlephilhower** [reply]
Not really.  I'm trying to get a basic self contained sketch.  Right now only you can do anything with th sketch you posted so it's not possible to look at it running.

So, to isolate things and make it reproducible for others I was suggesting a simple sketch that puts a PWM @1KHz out on GP0 and does an IRQ on GP1.  Anyone can just jumper the GP0 to GP1 pins and reproduce the test and verify that either a) we're OK and do get 1KHZ IRQs or b) we're really messed up and only getting 100hz or less...

**Comment by pcfreak1201** [comment]
Jupp, its the 25kHz PWM what screwed it up:

```
#define IN1 5
#define PWM1  6
#define POTI  A2

volatile uint32_t u32_IN1 = 0;

void __not_in_flash_func(countOne)()
{
  u32_IN1++;
}

void setup(){
    Serial.begin(115200);

    analogReadResolution(10);   // 10bit ADC-Val
    analogWriteResolution(10);  // 10bit PWM-Res.
    analogWriteFreq(1000); // 1kHz
    analogWriteRange(1000); // 0-1000 "%"

    pinMode(IN1, INPUT); 
    pinMode(PWM1, OUTPUT);
    attachInterrupt(digitalPinToInterrupt(IN1), countOne, FALLING);
}

void loop(){
  
uint16_t  uiAdcIn = analogRead(POTI);

  analogWrite(PWM1, map(uiAdcIn, 2, 1021, 0, 1000));
  Serial.println(u32_IN1); // counts per loop
  u32_IN1 = 0;

  delay(100);
}
```

counts to 100-101 - but the 25kHz are needed for the fan-PWM. So what should I do?

**Comment by Andy2No** [comment, marked as answer]
@pcfreak1201 It looks like you're calling analogWrite() too often.  Each time you do that it may start the PWM cycle again; I'm not sure.  Either way, there's no benefit to updating more often than once in every several cycles of its output.  Have you checked it's producing what you think it is from the PWM pin?

If you're aiming to produce a 20Hz PWM signal, you probably shouldn't try updating the PWM value more than a few times a second.  If you want to time how many counts there were in a certain time interval, you should probably set that quite a bit slower than the update period, otherwise you won't be getting much precision in the answer - the count will always be very small, so difficult to compare differences.

Also bear in mind that the ADCs are quite noisy, and non-linear.  The value will keep changing even though you're not moving the pot, and a small movement won't necessarily always result in a small change in the output - there can be big jumps in it because the ADCs have quite serious design flaws.

**Comment by pcfreak1201** [comment]
Ahh, OK. I'll try to calculate as mich as possible in the loop and wait min time before wrting the new value. That could work. Otherwise I'll could use a state-machine to slow down. Thanks, I'll report ...

**Reply by pcfreak1201** [reply]
Confirmed! Working with a state-machine and setting PWM / reading ADC only once, do the trick!
